From 12920e885030f1b76e2dcb5258c466b24832c119 Mon Sep 17 00:00:00 2001
From: Thu Nguyen <thu@os.amperecomputing.com>
Date: Mon, 28 Jul 2025 09:54:13 +0000
Subject: [PATCH] Host `CredentialBootstrapping` at the user path

As [1] and [2], the added `CredentialBootstrapping` D-Bus interface
which includes `EnableAfterReset`, `Enabled` and `RoleId` will be hosted
by phophos-user-manager. The values of these properties in the interface
are updated by phosphor-ipmi-host when handling ipmi command `Get
bootstrap account credentials (NetFn 2Ch, Command 02h)`.
Update code to support that. The init values of `EnableAfterReset` and
`Enabled` will be set base on the configured option
`credential_bootstrapping_enable_after_reset'`. The `RoleId` of
bootStrap user will always be `User`.

[1] https://gerrit.openbmc.org/c/openbmc/docs/+/79327/19/designs/bootstrap-account-access-restriction.md#186
[2] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/81735

Tested:
```text
$busctl introspect xyz.openbmc_project.User.Manager /xyz/openbmc_project/user
NAME                                                      TYPE
SIGNATURE  RESULT/VALUE                             FLAGS
xyz.openbmc_project.HostInterface.CredentialBootstrapping interface - -                                        -
.EnableAfterReset                                         property  b true                                     emits-change
.Enabled                                                  property  b true                                     emits-change writable
.RoleId                                                   property  s "User"                                   emits-change writable
```

Upstream-Status: Pending

Change-Id: Ie70583bb1d17274b0c07e40320e84e4d871abb36
Signed-off-by: Thu Nguyen <thu@os.amperecomputing.com>
---
 meson.build   |  4 ++++
 meson.options |  7 +++++++
 user_mgr.cpp  | 22 +++++++++++++++++++---
 user_mgr.hpp  | 17 +++++++++++++----
 users.cpp     |  2 +-
 5 files changed, 44 insertions(+), 8 deletions(-)

diff --git a/meson.build b/meson.build
index 9b0040f..69f8b55 100644
--- a/meson.build
+++ b/meson.build
@@ -15,6 +15,10 @@ if get_option('root_user_mgmt').allowed()
     add_project_arguments('-DENABLE_ROOT_USER_MGMT', language: 'cpp')
 endif
 
+if get_option('credential_bootstrapping_enable_after_reset').allowed()
+    add_project_arguments('-DENABLE_BOOTSTRAP_AFTER_RESET', language: 'cpp')
+endif
+
 conf_data = configuration_data()
 
 conf_data.set_quoted(
diff --git a/meson.options b/meson.options
index 95df809..a9af87b 100644
--- a/meson.options
+++ b/meson.options
@@ -15,3 +15,10 @@ option(
     value: 64,
     description: 'Max password Length default value as per NIST guideline recommendation',
 )
+
+option(
+    'credential_bootstrapping_enable_after_reset',
+    type: 'feature',
+    description: 'Enable Credential bootstrapping after reset service or host',
+    value: 'enabled',
+)
diff --git a/user_mgr.cpp b/user_mgr.cpp
index 76c5d3e..2286dc6 100644
--- a/user_mgr.cpp
+++ b/user_mgr.cpp
@@ -31,6 +31,7 @@
 #include <unistd.h>
 
 #include <boost/algorithm/string/split.hpp>
+#include <boost/container/flat_map.hpp>
 #include <phosphor-logging/elog-errors.hpp>
 #include <phosphor-logging/elog.hpp>
 #include <phosphor-logging/lg2.hpp>
@@ -1451,6 +1452,19 @@ void UserMgr::initializeAccountPolicy()
     }
 }
 
+void UserMgr::initializeCredentialBootstraping()
+{
+#ifdef ENABLE_BOOTSTRAP_AFTER_RESET
+    HostInterface::enableAfterReset(true);
+    HostInterface::enabled(true);
+    HostInterface::roleId("ReadOnly");
+#else
+    HostInterface::enableAfterReset(false);
+    HostInterface::enabled(false);
+    HostInterface::roleId("ReadOnly");
+#endif
+}
+
 void UserMgr::initUserObjects(void)
 {
     // All user management lock has to be based on /etc/shadow
@@ -1529,7 +1543,7 @@ void UserMgr::load()
             .value_or(std::optional(MultiFactorAuthType::None));
     if (authType)
     {
-        enabled(*authType, true);
+        multiFactorAuthEnabled(*authType, true);
     }
 }
 
@@ -1545,6 +1559,7 @@ UserMgr::UserMgr(sdbusplus::bus_t& bus, const char* path) :
     std::sort(groupsMgr.begin(), groupsMgr.end());
     UserMgrIface::allGroups(groupsMgr);
     initializeAccountPolicy();
+    initializeCredentialBootstraping();
     load();
     initUserObjects();
     // emit the signal
@@ -1599,9 +1614,10 @@ std::vector<std::string> UserMgr::getFailedAttempt(const char* userName)
     return executeCmd("/usr/sbin/faillock", "--user", userName);
 }
 
-MultiFactorAuthType UserMgr::enabled(MultiFactorAuthType value, bool skipSignal)
+MultiFactorAuthType UserMgr::multiFactorAuthEnabled(MultiFactorAuthType value,
+                                                    bool skipSignal)
 {
-    if (value == enabled())
+    if (value == multiFactorAuthEnabled())
     {
         return value;
     }
diff --git a/user_mgr.hpp b/user_mgr.hpp
index 27c5d41..e156762 100644
--- a/user_mgr.hpp
+++ b/user_mgr.hpp
@@ -20,12 +20,15 @@
 #include <sys/wait.h>
 #include <unistd.h>
 
+#include <boost/process/v1/child.hpp>
+#include <boost/process/v1/io.hpp>
 #include <phosphor-logging/elog-errors.hpp>
 #include <phosphor-logging/elog.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/bus.hpp>
 #include <sdbusplus/server/object.hpp>
 #include <xyz/openbmc_project/Common/error.hpp>
+#include <xyz/openbmc_project/HostInterface/CredentialBootstrapping/server.hpp>
 #include <xyz/openbmc_project/User/AccountPolicy/server.hpp>
 #include <xyz/openbmc_project/User/Manager/server.hpp>
 #include <xyz/openbmc_project/User/MultiFactorAuthConfiguration/server.hpp>
@@ -59,10 +62,11 @@ using MultiFactorAuthConfigurationIface =
     sdbusplus::xyz::openbmc_project::User::server::MultiFactorAuthConfiguration;
 
 using TOTPStateIface = sdbusplus::xyz::openbmc_project::User::server::TOTPState;
+using HostInterface = base::HostInterface::server::CredentialBootstrapping;
 
 using Ifaces = sdbusplus::server::object_t<UserMgrIface, AccountPolicyIface,
                                            MultiFactorAuthConfigurationIface,
-                                           TOTPStateIface>;
+                                           TOTPStateIface, HostInterface>;
 
 using Privilege = std::string;
 using GroupList = std::vector<std::string>;
@@ -351,12 +355,12 @@ class UserMgr : public Ifaces
     void createGroup(std::string groupName) override;
 
     void deleteGroup(std::string groupName) override;
-    MultiFactorAuthType enabled() const override
+    MultiFactorAuthType multiFactorAuthEnabled()
     {
         return MultiFactorAuthConfigurationIface::enabled();
     }
-    MultiFactorAuthType enabled(MultiFactorAuthType value,
-                                bool skipSignal) override;
+    MultiFactorAuthType multiFactorAuthEnabled(MultiFactorAuthType value,
+                                               bool skipSignal);
     bool secretKeyRequired(std::string userName) override;
     static std::vector<std::string> readAllGroupsOnSystem();
     void load();
@@ -504,6 +508,11 @@ class UserMgr : public Ifaces
 
     void initializeAccountPolicy();
 
+    /** @brief Init the `EnabledAfterReset`, `Enabled`, RoleId properties
+     *  in xyz.openbmc_project.HostInterface.CredentialBootstrapping
+     */
+    void initializeCredentialBootstraping();
+
     /** @brief checks if the group creation meets all constraints
      * @param groupName - group to check
      */
diff --git a/users.cpp b/users.cpp
index 9008008..12d6c22 100644
--- a/users.cpp
+++ b/users.cpp
@@ -225,7 +225,7 @@ bool changeFileOwnership(const std::string& filePath,
 }
 bool Users::checkMfaStatus() const
 {
-    return (manager.enabled() != MultiFactorAuthType::None &&
+    return (manager.multiFactorAuthEnabled() != MultiFactorAuthType::None &&
             Interfaces::bypassedProtocol() == MultiFactorAuthType::None);
 }
 std::string Users::createSecretKey()
-- 
2.34.1

